### ISSUE TITLE
Migrate tests to JUnit 5

### ISSUE THREAD

**Issue by rhusar** [arquillian/arquillian-extension-warp#220]
Investigate state of JUnit 5 support.

**Comment by WolfgangHG** [comment]
Main problem: the `org.jboss.arquillian.warp.impl.testutils.SeparatedClassloaderRunner` must be rewritten and converted to an Extension.

**Comment by WolfgangHG** [comment]
Seems JUnit5 does not support the scenario that a single test is run in a different classloader.
See this issue [Introduce extension API for providing a custom ClassLoader](/junit-team/junit5/issues/201), which is closed as the original reporter found a workaround.

In this issue, I found a link to https://junit.org/junit5/docs/snapshot/user-guide/#launcher-api-launcher-interceptors-custom which could be used to change the classloader. But as far as I understand it, this is done for the whole testsuite, while we would need it only for a few tests in the testsuite.

@rhusar Do you have any idea?

**Comment by WolfgangHG** [comment]
We could extract all tests that require the SeparatedClassloader approach to an extra project. In the first step, this old project could still be run with JUnit4. What do you think? 
If the LauncherInterceptor requires us to do different runs, it would also make it easier to handle "regular" JUnit runs and those special ones.

**Comment by rhusar** [comment]
> We could extract all tests that require the SeparatedClassloader approach to an extra project. In the first step, this old project could still be run with JUnit4. What do you think? If the LauncherInterceptor requires us to do different runs, it would also make it easier to handle "regular" JUnit runs and those special ones.

We do need to support both JUnit 4 and JUnit 5 within the same project branch. So that means we might need a new Warp module for JUnit 5 but definitely a new JUnit 5 surefire execution. Perhaps a common module that has the common JUnit classes.

**Comment by WolfgangHG** [comment]
Maybe I have misunderstanding here, but I think that JUnit is only used in the internal testsuite of warp. All dependencies have scope "test". There is no code of the extension itself that uses JUnit. So a user of the extension should be free to use JUnit4 or JUnit5. Or is our test dependency on JUnit propagated to all clients using the warp BOM?

**Comment by rhusar** [comment]
@WolfgangHG You might be right, I have not looked at this yet - the scope of the issue is really 'Investigate state of JUnit 5 support.' - as this is mostly implemented as an extension, then perhaps it just works? In that case, maybe a smoke test running with JUnit 5 would suffice.

**Comment by WolfgangHG** [comment]
@rhusar I gave my small test project (still JavaEE8, references Warp 1.1.0) a try and converted it to JUnit 5 - no problems in this simple test.
If you want to take a look: attached is the project: [StatelessMaven.zip](https://github.com/user-attachments/files/17997019/StatelessMaven.zip)

The test class is found at "/StatelessMaven-web/src/test/java/de/hsrm/cs/javaee8/statelessmaven/web/test/WarpIT.java". 
"StatelessMaven-web/pom.xml" and "StatelessMaven-ejb/pom.xml" both declare JUnit5 dependencies.
The project uses the bom "org.wildfly.bom:wildfly-jakartaee8-with-tools:26.0.0.Final", which provides a dependency on JUnit4. But I hope this does not conflict here.

**Comment by WolfgangHG** [comment]
@rhusar I tried to convert the test suite to JUnit 5, code can be found in my fork: https://github.com/WolfgangHG/arquillian-extension-warp/tree/junit5
In the first step, I commented all code that uses `org.jboss.arquillian.warp.impl.testutils.SeparatedClassloaderRunner`.

The projects "ftest" and "jsf-ftest" work again.

But there are several tests in the "impl" project that fail. The reason seems to be that injection does not happen.
Sample: https://github.com/WolfgangHG/arquillian-extension-warp/blob/junit5/impl/src/test/java/org/jboss/arquillian/warp/impl/client/eventbus/TestCommandEventBusLifecycle.java  => the injection for fields of arquillian type `Event<Before>` or  `Event<After>` does not happen, thus they are null. Do you have any idea? The test class is not annotated with the JUnit4 `@ArquillianRunner` or the JUnit5 `@ArquillianExtension`.

As you wrote that the scope of this issue is "investigate whether something needs to be done to support JUnit5 on the consumer side" (and my opinion is that it works without changes), I can create a new issue for converting the test suite.

**Comment by rhusar** [comment]
> @rhusar I tried to convert the test suite to JUnit 5, code can be found in my fork: https://github.com/WolfgangHG/arquillian-extension-warp/tree/junit5 In the first step, I commented all code that uses `org.jboss.arquillian.warp.impl.testutils.SeparatedClassloaderRunner`.

I am not up to speed why the SeparatedClassloaderRunner is needed especially since there is no javadoc :-) But I can have a look later.

> The projects "ftest" and "jsf-ftest" work again.

We can theoretically do these in parts too, i.e. per module.

> But there are several tests in the "impl" project that fail. The reason seems to be that injection does not happen. Sample: https://github.com/WolfgangHG/arquillian-extension-warp/blob/junit5/impl/src/test/java/org/jboss/arquillian/warp/impl/client/eventbus/TestCommandEventBusLifecycle.java => the injection for fields of arquillian type `Event<Before>` or `Event<After>` does not happen, thus they are null. Do you have any idea? The test class is not annotated with the JUnit4 `@ArquillianRunner` or the JUnit5 `@ArquillianExtension`.

There was problem with injecting to methods, but not to fields. I can have a look.

> As you wrote that the scope of this issue is "investigate whether something needs to be done to support JUnit5 on the consumer side" (and my opinion is that it works without changes), I can create a new issue for converting the test suite.

OK, that's good news! So the SeparatedClassloaderRunner is part of the project testing infrastructure - not something that users need to do.

We can keep using this issue for the upgrade, if we want to proceed with that.

**Comment by WolfgangHG** [comment]
Yes, [SeparatedClassLoaderRunner](https://github.com/arquillian/arquillian-extension-warp/blob/main/impl/src/test/java/org/jboss/arquillian/warp/impl/testutils/SeparatedClassloaderRunner.java) is in the "test" path, so I assume it is only used for internal tests and not public api.

My plan is to experiment with the JUnit5 `LauncherInterceptor` next week, which hopefully is the replacement for `SeparatedClassloaderRunner`

**Comment by WolfgangHG** [comment]
The LauncherInterceptor does not seem to work, as it is initialized only once per test suite. I asked a question in the JUnit5 repository: https://github.com/junit-team/junit5/discussions/4203

**Comment by WolfgangHG** [comment]
@rhusar while struggling with the SeparatedClassloaderRunner replacement suggested in my JUnit question, I got an idea why so many test in the "impl" project are failing (see e.g. the error for the test `org.jboss.arquillian.warp.impl.server.lifecycle.TestLifecycleTest` in https://github.com/WolfgangHG/arquillian-extension-warp/actions/runs/12337870415/job/34432140414#step:4:318 ): many tests in the "impl" project are indirect subclasses of `org.jboss.arquillian.core.test.AbstractManagerTestBase`, which uses JUnit4 annotations like `@org.junit.Before` instead of the Jupiter annotations, and several times `@org.junit.Assert`.

Is there a replacement in the JUnit5 module? Sounds like a "simple" copy of the class (and its subclasses) to https://github.com/arquillian/arquillian-core/tree/main/junit5 - or some smarter approach ;-).

**Comment by WolfgangHG** [comment]
@rhusar I have a working replacement for SeparatedClassloaderRunner based on the suggestion from https://github.com/junit-team/junit5/discussions/4203, but this one causes conflicts and errors with other tests that tamper with classloaders (e.g. [TestMigratedInspection.java](https://github.com/WolfgangHG/arquillian-extension-warp/blob/junit5/impl/src/test/java/org/jboss/arquillian/warp/impl/client/transformation/TestMigratedInspection.java) ).
As I don't find a resolution for this, the only solution that I can think is to extract all SeparatedClassloaderRunner to an extra project.
What do you think about this? Or should we stick to JUnit4 in the test suite?

About the second problem with `AbstractManagerTestBase`: I created https://www.github.com/arquillian/arquillian-core/issues/665. Hopefully, someone from the arquillian team can comment on this.

**Comment by rhusar** [comment]
> @rhusar I have a working replacement for SeparatedClassloaderRunner based on the suggestion from https://github.com/junit-team/junit5/discussions/4203, but this one causes conflicts and errors with other tests that tamper with classloaders (e.g. [TestMigratedInspection.java](https://github.com/WolfgangHG/arquillian-extension-warp/blob/junit5/impl/src/test/java/org/jboss/arquillian/warp/impl/client/transformation/TestMigratedInspection.java) ).

That looks promising.

> As I don't find a resolution for this, the only solution that I can think is to extract all SeparatedClassloaderRunner to an extra project.

You mean module right. Yes, that will work too.

> What do you think about this? Or should we stick to JUnit4 in the test suite?

We could just migrate module by module, doesn't have to be done in one take. 

> About the second problem with AbstractManagerTestBase: I created https://www.github.com/arquillian/arquillian-core/issues/665. Hopefully, someone from the arquillian team can comment on this.

Um, I am not sure why tests would be extending AbstractManagerTestBase, they should really just use `@ExtendWith(ArquillianExtension.class)` instead.

**Comment by WolfgangHG** [comment]
As far as I understand it, `@ExtendWith(ArquillianExtension.class)` makes test deployable on a server. All tests in the impl project just run locally, without being deployed to a server. Or do I have a misunderstanding?

**Comment by rhusar** [comment]
> As far as I understand it, `@ExtendWith(ArquillianExtension.class)` makes test deployable on a server. All tests in the impl project just run locally, without being deployed to a server. Or do I have a misunderstanding?

OK I see what you mean, while the `impl` module is not run in a container, this runner does not require deployments or anything. Just couple `@ExtendWith(ArquillianExtension.class)` with `@org.jboss.arquillian.container.test.api.RunAsClient`. I am not sure if that gives our everything that the Warp tests need, but this does run a test with ARQ also without a container.

**Comment by WolfgangHG** [comment]
@rhusar I pushed the SeparateClassLoader tests to a new project "impl_test_separatecl". Could you take a look at the commit and give me feedback?
https://github.com/WolfgangHG/arquillian-extension-warp/commit/44cb5984e7fd9d16e03643a3cd3cbd7222719942

**Comment by rhusar** [comment]
> [@rhusar](https://github.com/rhusar) I pushed the SeparateClassLoader tests to a new project "impl_test_separatecl". Could you take a look at the commit and give me feedback? [WolfgangHG@44cb598](https://github.com/WolfgangHG/arquillian-extension-warp/commit/44cb5984e7fd9d16e03643a3cd3cbd7222719942)

You can submit a draft pull request instead, that's better 👍🏼

**Comment by WolfgangHG** [comment]
I wanted to get a first review before continuing to work on the "AbstractManagerTestBase" problem ;-).

I fear I broke my [junit5](https://github.com/WolfgangHG/arquillian-extension-warp/tree/junit5) branch. I did a "git pull --rebase upstream main" and then "git push origin --force" (as you suggested some time ago), and after resolving some merge conflicts, github tells me "This branch is 11 commits ahead of, 13 commits behind arquillian/arquillian-extension-warp:main." Do you have any hints how to resolve this? I will give up for today....

**Comment by WolfgangHG** [comment]
@rhusar About `AbstractManagerTestBase` and replacing the base class with `@ExtendWith(ArquillianExtension.class)` and `@RunAsClient`: this does not work e.g. for [TestLifecycleTest](https://github.com/arquillian/arquillian-extension-warp/blob/main/impl/src/test/java/org/jboss/arquillian/warp/impl/server/lifecycle/TestLifecycleTest.java), as this one uses methods of `AbstractManagerTestBase` ("fire", "assertEventFired").
I copied `AbstractManagerTestBase`, `AbstractTestTestBase` and `AbstractContainerTestTestBase` from arquillian to warp and changed the JUnit annotations. After this, the `TestLifecycleTest` worked.
So this comes back to https://www.github.com/arquillian/arquillian-core/issues/665 I think.

Unfortunately, the test suite is flooded with mockito errors now:

```
[ERROR] org.jboss.arquillian.warp.impl.client.filter.http.TestHttpFilters.testShouldMatchUri -- Time elapsed: 0.873 s <<< ERROR!
org.mockito.exceptions.misusing.UnnecessaryStubbingException:

Unnecessary stubbings detected.
Clean & maintainable test code requires zero unnecessary code.
Following stubbings are unnecessary (click to navigate to relevant line of code):
  1. -> at org.jboss.arquillian.warp.impl.client.filter.http.TestHttpFilters.createRequest(TestHttpFilters.java:410)
  2. -> at org.jboss.arquillian.warp.impl.client.filter.http.TestHttpFilters.createRequest(TestHttpFilters.java:410)
Please remove unnecessary stubbings or use 'lenient' strictness. More info: javadoc for UnnecessaryStubbingException class.
        at org.mockito.junit.jupiter.MockitoExtension.afterEach(MockitoExtension.java:197)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
```

I think it could be e.g. this issue: https://www.github.com/mockito/mockito/issues/1540
Still needs further research whether it is actually a mockito issue and whether further "lenient" calls are required and reasonable.

**Comment by WolfgangHG** [comment]
About the mockito lenient errors, e.g. in [TestHttpFilters.java](https://github.com/WolfgangHG/arquillian-extension-warp/blob/junit5/impl/src/test/java/org/jboss/arquillian/warp/impl/client/filter/http/TestHttpFilters.java): two mocked variables `request` and `invalidRequest` are initialized in a `@BeforeEach` method, but not all tests use them or call all of their mocked functions.
It seems the `MockitoJUnitRunner` behaves differently than the `MockitoExtension`. Apparently, the runner detects that mock objects are accessed by **any** test, while the extension validates this for **each** test. With diagnostics output, I could verify that there is no test in TestHttpFilters.java that indirectly invokes all four stubs that are created in TestHttpFilters.createRequest, but all of them are invoked by some of the test methods.
So I am quite sure that it is the same problem like https://www.github.com/mockito/mockito/issues/1540 

For this test, it would be sufficient to create the request variables only once (a `@BeforeAll` annotated method). But Mockito does not support those "before class" mocks (see comments in linked issue) and does not report unnessary stubbing warnings at all.

To sum this up: I think easiest workaround would be to mark all stubs that cause those errors with `lenient()`. Not nice, but mockito changed its behavior.

There are other tests that fail, but hopefully it is the same issue.
